OpenRA.Mods.Common/Traits/World/PathFinder.cs

```diff
@@ -25,166 +27,95 @@ public override object Create(ActorInitializer init)
-		/// Returned path is *reversed* and given target to source.
-		/// </summary>
-		List<CPos> FindUnitPath(CPos source, CPos target, Actor self, Actor ignoreActor, BlockedByActor check);
-
-		/// <summary>
-		/// Expands the path search until a path is found, and returns that path.
-		/// Returned path is *reversed* and given target to source.
-		/// </summary>
-		List<CPos> FindPath(PathSearch search);
-
-		/// <summary>
-		/// Expands both path searches until they intersect, and returns the path.
-		/// Returned path is from the source of the first search to the source of the second search.
-		/// </summary>
-		List<CPos> FindBidiPath(PathSearch fromSrc, PathSearch fromDest);
-	}
-
 	public class PathFinder : IPathFinder
 	{
 		public static readonly List<CPos> NoPath = new List<CPos>(0);
 
 		readonly World world;
-		DomainIndex domainIndex;
-		bool cached;
 
 		public PathFinder(World world)
 		{
 			this.world = world;
 		}
 
 		/// <summary>
-		/// Calculates a path for the actor from source to target.
+		/// Calculates a path for the actor from multiple possible sources to target.
 		/// Returned path is *reversed* and given target to source.
+		/// The shortest path between a source and the target is returned.
 		/// </summary>
-		public List<CPos> FindUnitPath(CPos source, CPos target, Actor self, Actor ignoreActor, BlockedByActor check)
+		/// <remarks>
+		/// Searches that provide a multiple source cells are slower than those than provide only a single source cell,
+		/// as optimizations are possible for the single source case. Use searches from multiple source cells
+		/// sparingly.
+		/// </remarks>
+		public List<CPos> FindUnitPathToTargetCell(
+			Actor self, IEnumerable<CPos> sources, CPos target, BlockedByActor check,
+			Func<CPos, int> customCost = null,
+			Actor ignoreActor = null,
+			bool laneBias = true)
 		{
-			// PERF: Because we can be sure that OccupiesSpace is Mobile here, we can save some performance by avoiding querying for the trait.
-			var locomotor = ((Mobile)self.OccupiesSpace).Locomotor;
-
-			if (!cached)
-			{
-				domainIndex = world.WorldActor.TraitOrDefault<DomainIndex>();
-				cached = true;
-			}
+			var sourcesList = sources.ToList();
+			if (sourcesList.Count == 0)
+				throw new ArgumentException($"{nameof(sources)} must not be empty.", nameof(sources));
 
-			// If a water-land transition is required, bail early
-			if (domainIndex != null && !domainIndex.IsPassable(source, target, locomotor))
-				return NoPath;
+			var locomotor = GetLocomotor(self);
 
-			var distance = source - target;
-			var canMoveFreely = locomotor.CanMoveFreelyInto(self, target, check, null);
-			if (distance.LengthSquared < 3 && !canMoveFreely)
+			// If the target cell is inaccessible, bail early.
+			var inaccessible =
+				!locomotor.CanMoveFreelyInto(self, target, check, ignoreActor) ||
+				(!(customCost is null) && customCost(target) == PathGraph.PathCostForInvalidPath);
```

Comment:
Can we use `is not null` here?

Reply:
`Feature 'not pattern' is not available in C# 7.3. Please use language version 9.0 or greater.`

Reply:
Ah, well nvm then.